elf2flt: fix for segfault on some ARM ELFs

mpilawa opened the pull request:
I believe a bug was introduced in commit [1], which was done to move the .ARM.exidx input section from .data to .text output section.  However, in doing so, the dynamic memory allocation for .text [elf2flt.c:~L1907: 'text = xmalloc(text_len);'] was not modified to allow for the additional .ARM.exidx section.  The result was that most of the time malloc() allocated enough extra memory [due to page-size or similar allocation boundaries] such that a small additional section went unnoticed.  However, in unlucky cases, the memory required for just the .text section fit almost perfectly within an allocation boundary, and the extra .ARM.exidx section exceeded the allocation, and thus produced a segfault when the memory was written.

The fix here modifies the calculation of 'text_len' in main() with logic similar to that of the original fix in [1] to output_relocs(), such that the correct amount of memory is allocated.  The code is also modified such that 'data_len' is also calculated correctly, and does not over-allocate memory.  This is necessary because the logic in main() is not a single grouping of if-elseif... priority logic as in output_relocs().

The change also attempts to be even more specific with input section selections for .text and .data output sections.
.text is only selected for input sections flagged as (SEC_CODE || (SEC_DATA && SEC_READONLY && SEC_RELOC))
.data is only selected for input sections flagged as (SEC_DATA && !(SEC_READONLY && SEC_RELOC))

The change appears to work correctly for previously segfault-causing ELF with these processed sections...
SEC_FLAGS:0x0000011f SEC_NAME:.text
SEC_FLAGS:0x0000012f SEC_NAME:.ARM.exidx
SEC_FLAGS:0x00000127 SEC_NAME:.data
SEC_FLAGS:0x00000123 SEC_NAME:.tm_clone_table
SEC_FLAGS:0x0000012b SEC_NAME:.eh_frame
SEC_FLAGS:0x00000001 SEC_NAME:.bss
SEC_FLAGS:0x00000100 SEC_NAME:.stack
SEC_FLAGS:0x01800108 SEC_NAME:.comment
SEC_FLAGS:0x00000108 SEC_NAME:.ARM.attributes
SEC_FLAGS:0x0000210c SEC_NAME:.debug_aranges
SEC_FLAGS:0x0000210c SEC_NAME:.debug_info
SEC_FLAGS:0x00002108 SEC_NAME:.debug_abbrev
SEC_FLAGS:0x0000210c SEC_NAME:.debug_line
SEC_FLAGS:0x0000210c SEC_NAME:.debug_frame
SEC_FLAGS:0x01802108 SEC_NAME:.debug_str
SEC_FLAGS:0x0000210c SEC_NAME:.debug_loc
SEC_FLAGS:0x00002108 SEC_NAME:.debug_ranges

It should also be pointed out that this commit may impact the regression noted in [2] and pull-request [3].
With this code change in place, elf2flt will put section .eh_frame with flags=0x12b in the .data output section.  If the flags for an .eh_frame section were 0x12f (same as .ARM.exidx), then it would end up in .text output section.

A few cosmetic changes are included as well.

[1] https://github.com/uclinux-dev/elf2flt/commit/73325b7f209e0f68887333385184af275531427d
[2] https://github.com/uclinux-dev/elf2flt/issues/12
[3] https://github.com/uclinux-dev/elf2flt/pull/16

Signed-off-by: Mike Pilawa <Mike.Pilawa@csiro.au>

vapier commented on elf2flt.c:
```diff
@@ -1892,9 +1895,9 @@ int main(int argc, char *argv[])
     if (sec_vma < *vma) {
       if (*len > 0)
-	*len += sec_vma - *vma;
+        *len += sec_vma - *vma;
```
let's leave the whitespace alone in here.  this file is already a mismash of tabs-vs-spaces, so no sense in tweaking this one location.

mpilawa replied:
Okay, no worries.  Whitespace change removed in new force-pushed commit.

vapier commented on elf2flt.c:
```diff
@@ -1936,7 +1941,9 @@ int main(int argc, char *argv[])
   for (s = abs_bfd->sections; s != NULL; s = s->next)
-    if (s->flags & SEC_DATA) 
+    if ((s->flags & SEC_DATA) &&
+       ((s->flags & (SEC_READONLY | SEC_RELOC)) !=
```
i think i'm following why this is correct, but since the code is a bit nuanced (which is an understatement for this entire code base), can you update the comment here to add a bit more detail ?  maybe something like `/* Read in all data sections that haven't already been collected.  */` ?

mpilawa replied:
Okay, comments in code expanded a bit in new force-pushed commit.

mpilawa committed:
elf2flt: fix for segfault on some ARM ELFs

I believe a bug was introduced in commit [1], which was done to move the .ARM.exidx input section from .data to .text output section.  However, in doing so, the dynamic memory allocation for .text [elf2flt.c:~L1907: 'text = xmalloc(text_len);'] was not modified to allow for the additional .ARM.exidx section.  The result was that most of the time malloc() allocated enough extra memory [due to page-size or similar allocation boundaries] such that a small additional section went unnoticed.  However, in unlucky cases, the memory required for just the .text section fit almost perfectly within an allocation boundary, and the extra .ARM.exidx section exceeded the allocation, and thus produced a segfault when the memory was written.

The fix here modifies the calculation of 'text_len' in main() with logic similar to that of the original fix in [1] to output_relocs(), such that the correct amount of memory is allocated.  The code is also modified such that 'data_len' is also calculated correctly, and does not over-allocate memory.  This is necessary because the logic in main() is not a single grouping of if-elseif... priority logic as in output_relocs().

The change also attempts to be even more specific with input section selections for .text and .data output sections.
.text is only selected for input sections flagged as (SEC_CODE || (SEC_DATA && SEC_READONLY && SEC_RELOC))
.data is only selected for input sections flagged as (SEC_DATA && !(SEC_READONLY && SEC_RELOC))

The change appears to work correctly for previously segfault-causing ELF with these processed sections...
SEC_FLAGS:0x0000011f SEC_NAME:.text
SEC_FLAGS:0x0000012f SEC_NAME:.ARM.exidx
SEC_FLAGS:0x00000127 SEC_NAME:.data
SEC_FLAGS:0x00000123 SEC_NAME:.tm_clone_table
SEC_FLAGS:0x0000012b SEC_NAME:.eh_frame
SEC_FLAGS:0x00000001 SEC_NAME:.bss
SEC_FLAGS:0x00000100 SEC_NAME:.stack
SEC_FLAGS:0x01800108 SEC_NAME:.comment
SEC_FLAGS:0x00000108 SEC_NAME:.ARM.attributes
SEC_FLAGS:0x0000210c SEC_NAME:.debug_aranges
SEC_FLAGS:0x0000210c SEC_NAME:.debug_info
SEC_FLAGS:0x00002108 SEC_NAME:.debug_abbrev
SEC_FLAGS:0x0000210c SEC_NAME:.debug_line
SEC_FLAGS:0x0000210c SEC_NAME:.debug_frame
SEC_FLAGS:0x01802108 SEC_NAME:.debug_str
SEC_FLAGS:0x0000210c SEC_NAME:.debug_loc
SEC_FLAGS:0x00002108 SEC_NAME:.debug_ranges

It should also be pointed out that this commit may impact the regression noted in [2] and pull-request [3].
With this code change in place, elf2flt will put section .eh_frame with flags=0x12b in the .data output section.  If the flags for an .eh_frame section were 0x12f (same as .ARM.exidx), then it would end up in .text output section.

A few cosmetic changes are included as well.

[1] https://github.com/uclinux-dev/elf2flt/commit/73325b7f209e0f68887333385184af275531427d
[2] https://github.com/uclinux-dev/elf2flt/issues/12
[3] https://github.com/uclinux-dev/elf2flt/pull/16

Signed-off-by: Mike Pilawa <Mike.Pilawa@csiro.au>

vapier commented:
latest version looks great, but you'll need to rebase onto the latest branch.  can you pull+rebase so i can merge ?

mpilawa commented:
Should be all in sync and good to go now.

vapier commented:
thanks!
